pkg/controllers/dashboard/helm/repo_oci.go

```diff
@@ -286,3 +336,76 @@ func readBytes(configMapCache corev1controllers.ConfigMapClient, cm *corev1.Conf
 
 	return bytes, nil
 }
+
+// calculateBackoff gets the amount of time to wait for the next call.
+// Reference: https://github.com/oras-project/oras-go/blob/main/registry/remote/retry/policy.go#L95
+func calculateBackoff(clusterRepo *catalog.ClusterRepo, policy retryPolicy) time.Duration {
+	var h maphash.Hash
+	h.SetSeed(maphash.MakeSeed())
+	rand := rand.New(rand.NewSource(int64(h.Sum64())))
+	temp := float64(policy.MinWait) * math.Pow(2, float64(clusterRepo.Status.NumberOfRetries))
+	backoff := time.Duration(temp*(1-0.2)) + time.Duration(rand.Int63n(int64(2*0.2*temp)))
+	if backoff < policy.MinWait {
+		return policy.MinWait
+	}
+	if backoff > policy.MaxWait {
+		return policy.MaxWait
+	}
+	return backoff
+}
+
+// getRetryPolicy returns the retry policy for the repository using the values present in the spec
+// or, if they aren't present, the default values
+func getRetryPolicy(clusterRepo *catalog.ClusterRepo) retryPolicy {
+	// Default Values for exponentialBackOff function which is used
+	// to retry an HTTP call when 429 response code is hit.
+	var retryPolicy = retryPolicy{
+		MinWait:  1 * time.Second,
+		MaxWait:  5 * time.Second,
+		MaxRetry: 5,
+	}
+
+	if clusterRepo.Spec.ExponentialBackOffValues != nil {
+		if clusterRepo.Spec.ExponentialBackOffValues.MaxRetries > 0 {
+			retryPolicy.MaxRetry = clusterRepo.Spec.ExponentialBackOffValues.MaxRetries
+		}
+		if clusterRepo.Spec.ExponentialBackOffValues.MaxWait != "" {
+			maxWait, err := time.ParseDuration(clusterRepo.Spec.ExponentialBackOffValues.MaxWait)
+			if err == nil {
+				retryPolicy.MaxWait = maxWait
+			}
+		}
+		if clusterRepo.Spec.ExponentialBackOffValues.MinWait != "" {
+			minWait, err := time.ParseDuration(clusterRepo.Spec.ExponentialBackOffValues.MinWait)
+			if err == nil {
+				retryPolicy.MinWait = minWait
+			}
+		}
+	}
+	return retryPolicy
+}
+
+// shouldResetRetries checks to see if the interval has passed or if we need to do a force update
+func shouldResetRetries(clusterRepo *catalog.ClusterRepo) bool {
+	var lastStatusUpdate time.Time
+	for _, field := range clusterRepo.ManagedFields {
+		if field.Operation == metav1.ManagedFieldsOperationUpdate && field.Subresource == "status" {
+			lastStatusUpdate = field.Time.Time.UTC()
+		}
+	}
+	ociDownloaded := condition.Cond(catalog.OCIDownloaded)
+	ociDownloadedTime, _ := time.Parse(time.RFC3339, ociDownloaded.GetLastUpdated(clusterRepo))
+	// interval has passed. lastStatusUpdate will always be greater than zero if the ociDownloaded Condition exists
+	if !ociDownloadedTime.IsZero() && ociDownloadedTime.Add(interval).Before(timeNow().UTC()) && ociDownloadedTime.Add(interval).After(lastStatusUpdate) {
+		return true
+	}
+	// check for forceUpdate
+	if clusterRepo.Spec.ForceUpdate != nil && !lastStatusUpdate.IsZero() && clusterRepo.Spec.ForceUpdate.After(lastStatusUpdate) {
+		return true
+	}
+	// check if generation has changed
+	if clusterRepo.Generation > clusterRepo.Status.ObservedGeneration {
+		return true
+	}
```

Comment:
This is self-contained and could be the first check within func, which would short-circuit and avoid other unnecessary processing that happens before.